Litmusbook for couchbase server

deepav16 opened the pull request:
Signed-off-by: deepav16 <deepa.varambally@mayadata.io>

<!--  Thanks for sending a pull request!  Here are some tips for you -->

**What this PR does / why we need it**:

- This PR consist of  couchbase litmus job in order to deploy couchbase server 


**Special notes for your reviewer**:
@gprasath @nsathyaseelan @dargasudarshan

gprasath commented on apps/couchbase/deployers/couch_statefulset.yml:
```diff
@@ -0,0 +1,36 @@
+  name: couchbase
+spec:
+  serviceName: "couchbase"
+  replicas: 2
```
Can we get the number of replicas from user through env and update here

gprasath commented on apps/couchbase/deployers/couch_statefulset.yml:
```diff
@@ -0,0 +1,36 @@
+  - metadata:
+      name: couchbase-data
+      annotations:
+        volume.beta.kubernetes.io/storage-class: openebs-cstor-sparse
```
Replace storage class here with the corresponding env

gprasath replied:
use 'testclass' and replace as in 'pre_create_app_deploy.yml' util

gprasath commented on apps/couchbase/deployers/couch_statefulset.yml:
```diff
@@ -0,0 +1,36 @@
+            value: "false"
+  volumeClaimTemplates:
+  - metadata:
+      name: couchbase-data
```
provide placeholder for pvc

gprasath commented:
@deepav16 , can you provide the ansibletest output and the list of resources created.

deepav16 commented:
> @deepav16 , can you provide the ansibletest output and the list of resources created.
 


Ansibletest output:

-> kubectl logs litmus-couchbase-gkksp-r84st -n litmus

No config file found; using defaults
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
 [WARNING]: Found variable using reserved name: action

PLAY [localhost] ***************************************************************
2019-01-24T06:15:52.910734 (delta: 0.069137)         elapsed: 0.069137 ********
===============================================================================

TASK [Gathering Facts] *********************************************************
2019-01-24T06:15:52.927682 (delta: 0.016912)         elapsed: 0.086085 ********
ok: [127.0.0.1]

TASK [include_tasks] ***********************************************************
2019-01-24T06:15:58.861732 (delta: 5.933999)         elapsed: 6.020135 ********
included: /common/utils/create_testname.yml for 127.0.0.1

TASK [Record test instance/run ID] *********************************************
2019-01-24T06:15:58.964903 (delta: 0.103131)         elapsed: 6.123306 ********
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Construct testname appended with runID] **********************************
2019-01-24T06:15:59.022953 (delta: 0.057975)         elapsed: 6.181356 ********
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [include_tasks] ***********************************************************
2019-01-24T06:15:59.068399 (delta: 0.04539)         elapsed: 6.226802 *********
included: /common/utils/update_litmus_result_resource.yml for 127.0.0.1
TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
2019-01-24T06:15:59.163279 (delta: 0.094839)         elapsed: 6.321682 ********
changed: [127.0.0.1] => {"changed": true, "checksum": "716ed1b5d4c422f25d71fbf666e93e837926b5e6", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "7d8101f221c1cafc7895cbf001a9615a", "mode": "0644", "owner": "root", "size": 421, "src": "/root/.ansible/tmp/ansible-tmp-1548310559.23-242229829953995/source", "state": "file", "uid": 0}

TASK [Analyze the cr yaml] *****************************************************
2019-01-24T06:16:00.134665 (delta: 0.971338)         elapsed: 7.293068 ********
changed: [127.0.0.1] => {"changed": true, "cmd": "cat litmus-result.yaml", "delta": "0:00:00.529628", "end": "2019-01-24 06:16:01.232839", "rc": 0, "start": "2019-01-24 06:16:00.703211", "stderr": "", "stderr_lines": [], "stdout": "---\napiVersion: litmus.io/v1alpha1\nkind: LitmusResult\nmetadata:\n\n  # name of the litmus testcase\n  name: couchbase-deployment \nspec:\n\n  # holds information on the testcase\n  testMetadata:\n    app:  \n    chaostype:  \n\n  # holds the state of testcase,  manually updated by json merge patch\n  # result is the useful value today, but anticipate phase use in future \n  testStatus: \n    phase: in-progress  \n    result: none ", "stdout_lines": ["---", "apiVersion: litmus.io/v1alpha1", "kind: LitmusResult", "metadata:", "", "  # name of the litmus testcase", "  name: couchbase-deployment ", "spec:", "", "  # holds information on the testcase", "  testMetadata:", "    app:  ", "    chaostype:  ", "", "  # holds the state of testcase,  manually updated by json merge patch", "  # result is the useful value today, but anticipate phase use in future ", "  testStatus: ", "    phase: in-progress  ", "    result: none "]}

TASK [Apply the litmus result CR] **********************************************
2019-01-24T06:16:01.290826 (delta: 1.156106)         elapsed: 8.449229 ********
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.462507", "end": "2019-01-24 06:16:03.140503", "failed_when_result": false, "rc": 0, "start": "2019-01-24 06:16:01.677996", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/couchbase-deployment configured", "stdout_lines": ["litmusresult.litmus.io/couchbase-deployment configured"]}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
2019-01-24T06:16:03.198682 (delta: 1.907775)         elapsed: 10.357085 *******
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Analyze the cr yaml] *****************************************************
2019-01-24T06:16:03.257491 (delta: 0.058751)         elapsed: 10.415894 *******
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Apply the litmus result CR] **********************************************
2019-01-24T06:16:03.331236 (delta: 0.073697)         elapsed: 10.489639 *******
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [include_tasks] ***********************************************************
2019-01-24T06:16:03.461040 (delta: 0.129758)         elapsed: 10.619443 *******
included: /common/utils/pre_create_app_deploy.yml for 127.0.0.1

TASK [Check whether the provider storageclass is applied] **********************
2019-01-24T06:16:03.611917 (delta: 0.150683)         elapsed: 10.77032 ********
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get sc openebs-standalone", "delta": "0:00:00.741570", "end": "2019-01-24 06:16:04.650120", "rc": 0, "start": "2019-01-24 06:16:03.908550", "stderr": "", "stderr_lines": [], "stdout": "NAME                 PROVISIONER                    AGE\nopenebs-standalone   openebs .io/provisioner-iscsi   36d", "stdout_lines": ["NAME                 PROVISIONER                    AGE", "openebs-standalone   openebs.io/provisioner-iscsi   36d"]}

TASK [Replace the pvc placeholder with provider] *******************************
2019-01-24T06:16:04.754547 (delta: 1.14235)         elapsed: 11.91295 *********
changed: [127.0.0.1] => {"changed": true, "msg": "2 replacements made"}

TASK [Replace the storageclass placeholder with provider] **********************
2019-01-24T06:16:05.230553 (delta: 0.47596)         elapsed: 12.388956 ********
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [include_tasks] ***********************************************************
2019-01-24T06:16:05.611292 (delta: 0.38066)         elapsed: 12.769695 ********
included: /common/utils/fetch_replica_values.yml for 127.0.0.1

TASK [Get the application replica values from env] *****************************
2019-01-24T06:16:05.746105 (delta: 0.134768)         elapsed: 12.904508 *******
ok: [127.0.0.1] => {"ansible_facts": {"app_rkey": "replicas", "app_rvalue": "1"}, "changed": false}

TASK [Replace the application label placeholder] *******************************
2019-01-24T06:16:05.900772 (delta: 0.154621)         elapsed: 13.059175 *******
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Get the application label values from env] *******************************
2019-01-24T06:16:06.144089 (delta: 0.243207)         elapsed: 13.302492 *******
ok: [127.0.0.1] => {"ansible_facts": {"app_lkey": "app", "app_lvalue": "couchbase"}, "changed": false}

TASK [Replace the application label placeholder in deployment spec] ************
2019-01-24T06:16:06.253007 (delta: 0.108804)         elapsed: 13.41141 ********
changed: [127.0.0.1] => {"changed": true, "msg": "2 replacements made"}

TASK [Enable/Disable I/O based liveness probe] *********************************
2019-01-24T06:16:06.670324 (delta: 0.417145)         elapsed: 13.828727 *******
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [include_tasks] ***********************************************************
2019-01-24T06:16:06.803079 (delta: 0.132707)         elapsed: 13.961482 *******
included: /common/utils/create_ns.yml for 127.0.0.1

TASK [Obtain list of existing namespaces] **************************************
2019-01-24T06:16:06.892197 (delta: 0.089066)         elapsed: 14.0506 *********
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get ns --no-headers -o custom-columns=:metadata.name", "delta": "0:00:00.857026", "end": "2019-01-24 06:16:07.962341", "rc": 0, "start": "2019-01-24 06:16:07.105315", "stderr": "", "stderr_lines": [], "stdout": "app-esearch-ns\napp-percona-ns\ndefault\nkube-public\nkube-system\nlitmus\nmaya-system\nopenebs", "stdout_lines": ["app-esearch-ns", "app-percona-ns", "default", "kube-public", "kube-system", "litmus", "maya-system", "openebs"]}

TASK [Create test specific namespace.] *****************************************
2019-01-24T06:16:08.009188 (delta: 1.11694)         elapsed: 15.167591 ********
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl create ns app-couch-ns", "delta": "0:00:00.753159", "end": "2019-01-24 06:16:08.951813", "rc": 0, "start": "2019-01-24 06:16:08.198654", "stderr": "", "stderr_lines": [], "stdout": "namespace/app-couch-ns created", "stdout_lines": ["namespace/app-couch-ns created"]}

TASK [include_tasks] ***********************************************************
2019-01-24T06:16:09.010014 (delta: 1.00078)         elapsed: 16.168417 ********
included: /common/utils/status_testns.yml for 127.0.0.1

TASK [Checking the status  of test specific namespace.] ************************
2019-01-24T06:16:09.112302 (delta: 0.102199)         elapsed: 16.270705 *******
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get ns app-couch-ns -o custom-columns=:status.phase --no-headers", "delta": "0:00:00.911709", "end": "2019-01-24 06:16:10.195814", "rc": 0, "start": "2019-01-24 06:16:09.284105", "stderr": "", "stderr_lines": [], "stdout": "Active", "stdout_lines": ["Active"]}

TASK [include_tasks] ***********************************************************
2019-01-24T06:16:10.316626 (delta: 1.204214)         elapsed: 17.475029 *******
included: /common/utils/deploy_single_app.yml for 127.0.0.1

TASK [Deploying Couchbase] *****************************************************
2019-01-24T06:16:10.460288 (delta: 0.14357)         elapsed: 17.618691 ********
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f couch_statefulset.yml -n app-couch-ns", "delta": "0:00:00.960353", "end": "2019-01-24 06:16:11.795657", "rc": 0, "start": "2019-01-24 06:16:10.835304", "stderr": "", "stderr_lines": [], "stdout": "statefulset.apps/couchbase created", "stdout_lines": ["statefulset.apps/couchbase created"]}

TASK [include_tasks] ***********************************************************
2019-01-24T06:16:11.837776 (delta: 1.377396)         elapsed: 18.996179 *******
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [include_tasks] ***********************************************************
2019-01-24T06:16:11.899407 (delta: 0.061509)         elapsed: 19.05781 ********
included: /common/utils/check_replica_count.yml for 127.0.0.1

TASK [Obtain the number of replicas.] ******************************************
2019-01-24T06:16:11.972905 (delta: 0.07331)         elapsed: 19.131308 ********
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get statefulset -n app-couch-ns -l \"app=couchbase\" -o custom-columns=:spec.replicas", "delta": "0:00:00.865332", "end": "2019-01-24 06:16:13.002618", "rc": 0, "start": "2019-01-24 06:16:12.137286", "stderr": "", "stderr_lines": [], "stdout": "\n1", "stdout_lines": ["", "1"]}

TASK [Obtain the ready replica count and compare with the replica count.] ******
2019-01-24T06:16:13.045724 (delta: 1.072776)         elapsed: 20.204127 *******
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (30 retries left).
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (29 retries left).
changed: [127.0.0.1] => {"attempts": 3, "changed": true, "cmd": "kubectl get statefulset -n app-couch-ns -l \"app=couchbase\" -o custom-columns=:..readyReplicas", "delta": "0:00:00.828623", "end": "2019-01-24 06:18:16.876271", "rc": 0, "start": "2019-01-24 06:18:16.047648", "stderr": "", "stderr_lines": [], "stdout": "\n1", "stdout_lines": ["", "1"]}

TASK [Deprovisioning the Application] ******************************************
2019-01-24T06:18:16.934035 (delta: 123.888269)         elapsed: 144.092438 ****
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [set_fact] ****************************************************************
2019-01-24T06:18:16.993461 (delta: 0.059376)         elapsed: 144.151864 ******
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [include_tasks] ***********************************************************
2019-01-24T06:18:17.071884 (delta: 0.078312)         elapsed: 144.230287 ******
included: /common/utils/update_litmus_result_resource.yml for 127.0.0.1

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
2019-01-24T06:18:17.165337 (delta: 0.093313)         elapsed: 144.32374 *******
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Analyze the cr yaml] *****************************************************
2019-01-24T06:18:17.213454 (delta: 0.048)         elapsed: 144.371857 *********
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Apply the litmus result CR] **********************************************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Analyze the cr yaml] *****************************************************
2019-01-24T06:18:17.213454 (delta: 0.048)         elapsed: 144.371857 *********
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Apply the litmus result CR] **********************************************
2019-01-24T06:18:17.264287 (delta: 0.05076)         elapsed: 144.42269 ********
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
2019-01-24T06:18:17.316453 (delta: 0.05203)         elapsed: 144.474856 *******
changed: [127.0.0.1] => {"changed": true, "checksum": "6df6f9080e8feeef087982c8083905c70d41d217", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "af558e6976ea3c0d9fa1cec794573cf6", "mode": "0644", "owner": "root", "size": 419, "src": "/root/.ansible/tmp/ansible-tmp-1548310697.38-223189904505098/source", "state": "file", "uid": 0}

TASK [Analyze the cr yaml] *****************************************************
2019-01-24T06:18:18.961394 (delta: 1.644857)         elapsed: 146.119797 ******
changed: [127.0.0.1] => {"changed": true, "cmd": "cat litmus-result.yaml", "delta": "0:00:00.566984", "end": "2019-01-24 06:18:19.702333", "rc": 0, "start": "2019-01-24 06:18:19.135349", "stderr": "", "stderr_lines": [], "stdout": "---\napiVersion: litmus.io/v1alpha1\nkind: LitmusResult\nmetadata:\n\n  # name of the litmus testcase\n  name: couchbase-deployment \nspec:\n\n  # holds information on the testcase\n  testMetadata:\n    app:  \n    chaostype:  \n\n  # holds the state of testcase,  manually updated by json merge patch\n  # result is the useful value today, but anticipate phase use in future \n  testStatus: \n    phase: completed  \n    result: Pass ", "stdout_lines": ["---", "apiVersion: litmus.io/v1alpha1", "kind: LitmusResult", "metadata:", "", "  # name of the litmus testcase", "  name: couchbase-deployment ", "spec:", "", "  # holds information on the testcase", "  testMetadata:", "    app:  ", "    chaostype:  ", "", "  # holds the state of testcase,  manually updated by json merge patch", "  # result is the useful value today, but anticipate phase use in future ", "  testStatus: ", "    phase: completed  ", "    result: Pass "]}

TASK [Apply the litmus result CR] **********************************************
2019-01-24T06:18:19.830373 (delta: 0.868933)         elapsed: 146.988776 ******
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:00.890185", "end": "2019-01-24 06:18:20.935835", "failed_when_result": false, "rc": 0, "start": "2019-01-24 06:18:20.045650", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/couchbase-deployment configured", "stdout_lines": ["litmusresult.litmus.io/couchbase-deployment configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=30   changed=17   unreachable=0    failed=0

2019-01-24T06:18:20.972398 (delta: 1.141979)         elapsed: 148.130801 ******
===============================================================================

Resources created 

-> $kubectl get all -n app-couch-ns

NAME                                                                 READY     STATUS    RESTARTS   AGE
pod/couchbase-0                                                      1/1       Running   0          17m
pod/pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281-ctrl-56cb984c89-clxbz   2/2       Running   0          17m
pod/pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281-rep-6df4bb56c7-pfd8h    1/1       Running   0          17m

NAME                                                        TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                      AGE
service/pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281-ctrl-svc   ClusterIP   10.71.254.28   <none>        3260/TCP,9501/TCP,9500/TCP   17m

NAME                                                            DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281-ctrl   1         1         1            1           17m
deployment.apps/pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281-rep    1         1         1            1           17m

NAME                                                                       DESIRED   CURRENT   READY     AGE
replicaset.apps/pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281-ctrl-56cb984c89   1         1         1         17m
replicaset.apps/pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281-rep-6df4bb56c7    1         1         1         17m

NAME                         DESIRED   CURRENT   AGE
statefulset.apps/couchbase   1         1         17m

 -> $kubectl get pvc -n app-couch-ns

NAME                            STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS         AGE
openebs-couchbase-couchbase-0   Bound     pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281   5G         RWO            openebs-standalone   18m

->$ kubectl get pv -n app-couch-ns

NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS        CLAIM                                                    STORAGECLASS           REASON    AGE
pvc-8bd8fc8e-1f9f-11e9-bf34-42010a800281   5G         RWO            Delete           Bound         app-couch-ns/openebs-couchbase-couchbase-0               openebs-standalone               20m
pvc-ea5a1a22-1f9b-11e9-bf34-42010a800281   20G        RWO            Delete           Bound         app-esearch-ns/openebs-esearch-elasticsearch-logging-0   openebs-standard                 46m
